File: windows-notification-listener/TrafficManager.NotificationListener/SlackNotificationParser.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace TrafficManager.NotificationListener;

/// <summary>
/// Keep in sync with backend/app/integrations/slack_notification_parser.py
/// </summary>
internal static class SlackNotificationParser
{
    private static readonly Regex AggregateRe = new(
        @"^(?:\d+\s+new messages?|new activity(?: in slack)?|you have (?:unread |new )?messages?|unread messages?|new messages? in slack|slack$)\.?$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.Compiled);

    private static readonly Regex ChannelPrefixRe = new(
        @"^(?:channel:\s*|#)(.+)$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.Compiled);

    private static readonly Regex SenderBodyRe = new(
        @"^(.{1,80}?):\s+(.+)$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.Singleline | RegexOptions.Compiled);

    private static readonly Regex ThreadHintRe = new(
        @"\breplied to a thread\b|\bin a thread\b",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.Compiled);

    private static readonly Regex NonSlugRe = new(@"[^a-z0-9]+", RegexOptions.Compiled);

    public static ParseResult Parse(
        IReadOnlyList<string> textElements,
        AppIdentity identity,
        string? notificationId,
        string? createdAt,
        IEnumerable<string>? extraSourceIds = null,
        bool truncationFlagged = false)
    {
        var sourceKind = NotificationSourceClassifier.Classify(identity, extraSourceIds);
        var sourceId = NotificationSourceClassifier.SourceId(identity);
        var lines = Lines(textElements);
        var emptyId = MessageId(sourceId, notificationId, createdAt, null, "");
        if (sourceKind != NotificationSourceClassifier.SlackDesktop)
        {
            return Skip(sourceKind, sourceKind == NotificationSourceClassifier.Other ? "unrelated" : "browser_unknown", sourceId, emptyId);
        }

        if (lines.Count == 0)
        {
            return Skip(sourceKind, "empty", sourceId, emptyId);
        }

        if (lines.Any(IsAggregate) || IsAggregate(string.Join(" ", lines)))
        {
            return Skip(sourceKind, "aggregate", sourceId, emptyId);
        }

        var title = lines[0];
        var rest = lines.Skip(1).ToList();
        if (title.Equals("slack", StringComparison.OrdinalIgnoreCase) && rest.Count > 0)
        {
            title = rest[0];
            rest = rest.Skip(1).ToList();
            if (IsAggregate(title) || rest.Any(IsAggregate) || IsAggregate(string.Join(" ", new[] { title }.Concat(rest))))
            {
                return Skip(sourceKind, "aggregate", sourceId, emptyId);
            }
        }

        var conversationKind = "direct";
        string? conversationHint = null;
        string? senderName = null;
        var body = string.Join("\n", rest);
        var channelMatch = ChannelPrefixRe.Match(title);
        if (channelMatch.Success)
        {
            conversationKind = "channel";
            conversationHint = channelMatch.Groups[1].Value.Trim();
            var senderMatch = SenderBodyRe.Match(body);
            if (senderMatch.Success)
            {
                senderName = senderMatch.Groups[1].Value.Trim();
                body = senderMatch.Groups[2].Value.Trim();
            }
        }
        else if (rest.Count == 0 && SenderBodyRe.IsMatch(title))
        {
            var senderMatch = SenderBodyRe.Match(title);
            senderName = senderMatch.Groups[1].Value.Trim();
            body = senderMatch.Groups[2].Value.Trim();
            conversationHint = senderName;
        }
        else
        {
            senderName = title;
            conversationHint = title;
        }

        body = Normalize(body);
        if (string.IsNullOrWhiteSpace(body))
        {
            return Skip(sourceKind, "empty", sourceId, emptyId);
        }

        if (IsAggregate(body))
        {
            return Skip(sourceKind, "aggregate", sourceId, emptyId);
        }

        var hint = conversationHint ?? senderName ?? "unknown";
        var confidence = conversationKind == "channel" && !string.IsNullOrWhiteSpace(senderName)
            ? "high"
            : conversationKind == "direct" && !string.IsNullOrWhiteSpace(senderName)
                ? "medium"
                : "low";
        if (confidence == "low")
        {
            return Skip(sourceKind, "low_confidence", sourceId, emptyId);
        }

        var chatId = ChatId(conversationKind, hint);
        var messageId = MessageId(sourceId, notificationId, createdAt, senderName, body);
        return new ParseResult
        {
            SourceKind = sourceKind,
            SkipReason = null,
            SenderName = senderName,
            Text = body,
            ConversationHint = conversationHint,
            ConversationKind = conversationKind,
            MappingConfidence = confidence,
            IsTruncated = DetectTruncation(body, truncationFlagged),
            NotificationExternalId = messageId,
            ChatExternalId = chatId,
            ThreadHint = ThreadHintRe.IsMatch(string.Join("\n", lines)) ? "thread" : null,
            SourceId = sourceId,
        };
    }

    public static SlackNotificationEvent ToEvent(ParseResult parsed, string receivedAt)
    {
        return new SlackNotificationEvent
        {
            NotificationExternalId = parsed.NotificationExternalId,
            ReceivedAt = receivedAt,
            ConversationHint = parsed.ConversationHint,
            ConversationKind = parsed.ConversationKind,
            SenderName = parsed.SenderName,
            Text = parsed.Text ?? "",
            IsTruncated = parsed.IsTruncated,
            MappingConfidence = parsed.MappingConfidence,
            ThreadHint = parsed.ThreadHint,
            SourceId = parsed.SourceId,
        };
    }

    private static ParseResult Skip(string sourceKind, string reason, string sourceId, string emptyId)
    {
        return new ParseResult
        {
            SourceKind = sourceKind,
            SkipReason = reason,
            SourceId = sourceId,
            NotificationExternalId = emptyId,
        };
    }

    private static bool IsAggregate(string? value)
    {
        var text = Normalize(value);
        return text.Length > 0 && AggregateRe.IsMatch(text);
    }

    private static bool DetectTruncation(string text, bool flagged)
    {
        if (flagged)
        {
            return true;
        }

        var stripped = text.TrimEnd();
        return stripped.EndsWith("…", StringComparison.Ordinal) || stripped.EndsWith("...", StringComparison.Ordinal) || stripped.Length >= 220;
    }

    private static string ChatId(string kind, string hint)
    {
        var prefix = kind == "channel" ? "channel" : "dm";
        return $"notification:{prefix}:{Slug(hint)}";
    }

    private static string Slug(string value)
    {
        var lowered = Normalize(value).ToLowerInvariant().TrimStart('#');
        var slug = NonSlugRe.Replace(lowered, "-").Trim('-');
        if (slug.Length == 0)
        {
            slug = "unknown";
        }

        return slug.Length <= 80 ? slug : slug[..80];
    }

    private static string MessageId(string sourceId, string? notificationId, string? createdAt, string? sender, string text)
    {
        string material;
        if (!string.IsNullOrWhiteSpace(notificationId) && !string.IsNullOrWhiteSpace(sourceId))
        {
            material = $"{sourceId}\n{notificationId}\n{createdAt ?? ""}";
        }
        else
        {
            var bucket = (createdAt ?? "").Length >= 16 ? createdAt![..16] : createdAt ?? "";
            material = $"{sourceId}\n{sender ?? ""}\n{text}\n{bucket}";
        }

        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(material));
        var hex = Convert.ToHexString(hash).ToLowerInvariant()[..32];
        return "n_" + hex;
    }

    private static List<string> Lines(IEnumerable<string> textElements)
    {
        return textElements.Select(Normalize).Where(item => item.Length > 0).ToList();
    }

    private static string Normalize(string? value)
    {
        return Regex.Replace((value ?? "").Replace("\r\n", "\n"), "[ \t]+", " ").Trim();
    }
}

File: backend/app/integrations/slack_notification_parser.py
"""Parse Slack Desktop toast text into a local inbox event.

Never invents Slack channel IDs (C/D/G...), user IDs, or thread timestamps.
Keep in sync with windows-notification-listener SlackNotificationParser.cs.
"""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from datetime import datetime

from app.integrations.slack_notification_source import (
    BROWSER_UNKNOWN,
    OTHER,
    SLACK_DESKTOP,
    NotificationAppIdentity,
    classify_notification_source,
    source_id_for,
)

AGGREGATE_RE = re.compile(
    r"^(?:"
    r"\d+\s+new messages?|"
    r"new activity(?: in slack)?|"
    r"you have (?:unread |new )?messages?|"
    r"unread messages?|"
    r"new messages? in slack|"
    r"slack$"
    r")\.?$",
    re.IGNORECASE,
)
CHANNEL_PREFIX_RE = re.compile(r"^(?:channel:\s*|#)(.+)$", re.IGNORECASE)
SENDER_BODY_RE = re.compile(r"^(.{1,80}?):\s+(.+)$", re.DOTALL)
THREAD_HINT_RE = re.compile(r"\breplied to a thread\b|\bin a thread\b", re.IGNORECASE)
NON_SLUG_RE = re.compile(r"[^a-z0-9]+")
SLACK_API_ID_RE = re.compile(r"^[CDG][A-Z0-9]{8,}$")


@dataclass(frozen=True)
class SlackNotificationParseResult:
    source_kind: str
    skip_reason: str | None
    sender_name: str | None
    text: str | None
    conversation_hint: str | None
    conversation_kind: str
    mapping_confidence: str
    is_truncated: bool
    notification_external_id: str
    chat_external_id: str | None
    thread_hint: str | None
    source_id: str


def _normalize_line(value: str | None) -> str:
    return re.sub(r"[ \t]+", " ", str(value or "").replace("\r\n", "\n")).strip()


def _lines(text_elements: list[str] | tuple[str, ...]) -> list[str]:
    return [line for line in (_normalize_line(item) for item in text_elements) if line]


def is_aggregate_text(*parts: str | None) -> bool:
    for part in parts:
        value = _normalize_line(part)
        if value and AGGREGATE_RE.match(value):
            return True
    return False


def detect_truncation(text: str, *, flagged: bool = False) -> bool:
    if flagged:
        return True
    stripped = text.rstrip()
    if stripped.endswith("…") or stripped.endswith("..."):
        return True
    return len(stripped) >= 220


def slug_for(value: str) -> str:
    lowered = _normalize_line(value).lower().lstrip("#")
    slug = NON_SLUG_RE.sub("-", lowered).strip("-")
    return (slug or "unknown")[:80]


def notification_chat_id(*, kind: str, hint: str) -> str:
    prefix = "channel" if kind == "channel" else "dm"
    return f"notification:{prefix}:{slug_for(hint)}"


def notification_message_id(
    *,
    source_id: str,
    notification_id: str | None,
    created_at: str | None,
    sender_name: str | None,
    text: str,
) -> str:
    if notification_id and source_id:
        material = f"{source_id}\n{notification_id}\n{created_at or ''}"
    else:
        bucket = (created_at or "")[:16]
        material = f"{source_id}\n{sender_name or ''}\n{text}\n{bucket}"
    digest = hashlib.sha256(material.encode("utf-8")).hexdigest()[:32]
    return f"n_{digest}"


def _looks_like_slack_api_id(value: str | None) -> bool:
    return bool(value and SLACK_API_ID_RE.match(value))


def parse_slack_notification(
    text_elements: list[str] | tuple[str, ...],
    *,
    identity: NotificationAppIdentity,
    notification_id: str | None = None,
    created_at: str | None = None,
    extra_source_ids: tuple[str, ...] | list[str] = (),
    truncation_flagged: bool = False,
) -> SlackNotificationParseResult:
    source_kind = classify_notification_source(identity, extra_source_ids=extra_source_ids)
    source_id = source_id_for(identity)
    lines = _lines(text_elements)
    empty_id = notification_message_id(
        source_id=source_id,
        notification_id=notification_id,
        created_at=created_at,
        sender_name=None,
        text="",
    )
    if source_kind != SLACK_DESKTOP:
        reason = "unrelated" if source_kind == OTHER else "browser_unknown"
        return SlackNotificationParseResult(
            source_kind=source_kind,
            skip_reason=reason,
            sender_name=None,
            text=None,
            conversation_hint=None,
            conversation_kind="direct",
            mapping_confidence="low",
            is_truncated=False,
            notification_external_id=empty_id,
            chat_external_id=None,
            thread_hint=None,
            source_id=source_id,
        )
    if not lines:
        return SlackNotificationParseResult(
            source_kind=source_kind,
            skip_reason="empty",
            sender_name=None,
            text=None,
            conversation_hint=None,
            conversation_kind="direct",
            mapping_confidence="low",
            is_truncated=False,
            notification_external_id=empty_id,
            chat_external_id=None,
            thread_hint=None,
            source_id=source_id,
        )
    if any(is_aggregate_text(line) for line in lines) or is_aggregate_text(" ".join(lines)):
        return SlackNotificationParseResult(
            source_kind=source_kind,
            skip_reason="aggregate",
            sender_name=None,
            text=None,
            conversation_hint=None,
            conversation_kind="direct",
            mapping_confidence="low",
            is_truncated=False,
            notification_external_id=empty_id,
            chat_external_id=None,
            thread_hint=None,
            source_id=source_id,
        )

    title = lines[0]
    rest = lines[1:]
    if title.lower() == "slack" and rest:
        title = rest[0]
        rest = rest[1:]
        if is_aggregate_text(title, *rest, " ".join([title, *rest])):
            return SlackNotificationParseResult(
                source_kind=source_kind,
                skip_reason="aggregate",
                sender_name=None,
                text=None,
                conversation_hint=None,
                conversation_kind="direct",
                mapping_confidence="low",
                is_truncated=False,
                notification_external_id=empty_id,
                chat_external_id=None,
                thread_hint=None,
                source_id=source_id,
            )

    conversation_kind = "direct"
    conversation_hint: str | None = None
    sender_name: str | None = None
    body = "\n".join(rest)

    channel_match = CHANNEL_PREFIX_RE.match(title)
    if channel_match:
        conversation_kind = "channel"
        conversation_hint = channel_match.group(1).strip()
        sender_match = SENDER_BODY_RE.match(body)
        if sender_match:
            sender_name = sender_match.group(1).strip()
            body = sender_match.group(2).strip()
    elif not rest and SENDER_BODY_RE.match(title):
        sender_match = SENDER_BODY_RE.match(title)
        assert sender_match is not None
        sender_name = sender_match.group(1).strip()
        body = sender_match.group(2).strip()
        conversation_hint = sender_name
    else:
        sender_name = title
        conversation_hint = title
        if not body and rest:
            body = "\n".join(rest)

    body = _normalize_line(body)
    if not body:
        return SlackNotificationParseResult(
            source_kind=source_kind,
            skip_reason="empty",
            sender_name=sender_name,
            text=None,
            conversation_hint=conversation_hint,
            conversation_kind=conversation_kind,
            mapping_confidence="low",
            is_truncated=False,
            notification_external_id=empty_id,
            chat_external_id=None,
            thread_hint=None,
            source_id=source_id,
        )

    if is_aggregate_text(body):
        return SlackNotificationParseResult(
            source_kind=source_kind,
            skip_reason="aggregate",
            sender_name=sender_name,
            text=None,
            conversation_hint=conversation_hint,
            conversation_kind=conversation_kind,
            mapping_confidence="low",
            is_truncated=False,
            notification_external_id=empty_id,
            chat_external_id=None,
            thread_hint=None,
            source_id=source_id,
        )

    hint = conversation_hint or sender_name or "unknown"
    if conversation_kind == "channel" and sender_name and body:
        confidence = "high"
    elif conversation_kind == "direct" and sender_name and body:
        confidence = "medium"
    else:
        confidence = "low"

    if confidence == "low":
        return SlackNotificationParseResult(
            source_kind=source_kind,
            skip_reason="low_confidence",
            sender_name=sender_name,
            text=body,
            conversation_hint=conversation_hint,
            conversation_kind=conversation_kind,
            mapping_confidence=confidence,
            is_truncated=detect_truncation(body, flagged=truncation_flagged),
            notification_external_id=empty_id,
            chat_external_id=None,
            thread_hint=None,
            source_id=source_id,
        )

    chat_id = notification_chat_id(kind=conversation_kind, hint=hint)
    if _looks_like_slack_api_id(chat_id) or _looks_like_slack_api_id(hint):
        chat_id = notification_chat_id(kind=conversation_kind, hint=f"hint-{slug_for(hint)}")

    message_id = notification_message_id(
        source_id=source_id,
        notification_id=notification_id,
        created_at=created_at,
        sender_name=sender_name,
        text=body,
    )
    thread_hint = "thread" if THREAD_HINT_RE.search("\n".join(lines)) else None
    return SlackNotificationParseResult(
        source_kind=source_kind,
        skip_reason=None,
        sender_name=sender_name,
        text=body,
        conversation_hint=conversation_hint,
        conversation_kind=conversation_kind,
        mapping_confidence=confidence,
        is_truncated=detect_truncation(body, flagged=truncation_flagged),
        notification_external_id=message_id,
        chat_external_id=chat_id,
        thread_hint=thread_hint,
        source_id=source_id,
    )


def parse_created_at(value: str | None) -> datetime | None:
    if not value:
        return None
    text = value.strip()
    if text.endswith("Z"):
        text = text[:-1] + "+00:00"
    try:
        parsed = datetime.fromisoformat(text)
    except ValueError:
        return None
    return parsed
